syft/pkg/cataloger/javascript/parse_package_lock.go

```diff
@@ -47,38 +47,51 @@ func parsePackageLock(path string, reader io.Reader) ([]*pkg.Package, []artifact
 	var packages []*pkg.Package
 	dec := json.NewDecoder(reader)
 
+	var lock PackageLock
 	for {
-		var lock PackageLock
 		if err := dec.Decode(&lock); err == io.EOF {
 			break
 		} else if err != nil {
 			return nil, nil, fmt.Errorf("failed to parse package-lock.json file: %w", err)
 		}
-		licenseMap := make(map[string]string)
-		for _, pkgMeta := range lock.Packages {
-			var sb strings.Builder
-			sb.WriteString(pkgMeta.Resolved)
-			sb.WriteString(pkgMeta.Integrity)
-			licenseMap[sb.String()] = pkgMeta.License
-		}
+	}
 
+	if lock.LockfileVersion == 1 {
 		for name, pkgMeta := range lock.Dependencies {
-			var sb strings.Builder
-			sb.WriteString(pkgMeta.Resolved)
-			sb.WriteString(pkgMeta.Integrity)
-			var licenses []string
-			if license, exists := licenseMap[sb.String()]; exists {
-				licenses = append(licenses, license)
-			}
 			packages = append(packages, &pkg.Package{
 				Name:     name,
 				Version:  pkgMeta.Version,
 				Language: pkg.JavaScript,
 				Type:     pkg.NpmPkg,
-				Licenses: licenses,
 			})
 		}
 	}
 
+	if lock.LockfileVersion == 2 || lock.LockfileVersion == 3 {
+		for name, pkgMeta := range lock.Packages {
+			if name == "" {
+				continue
+			}
+
+			newPackage := &pkg.Package{
+				Name:     getNameFromPath(name),
```

Comment:
I think we might need to add a custom `NodeMetadata` or something to capture the full "name" so the license lookup will be accurate if there isn't a license defined in the `package-lock` file

Reply:
If we do that we should distinguish that from the existing `NpmPackageJsonMetadata` with something like `NpmPackageLockMetadata`.  If we go that directory it would be a good opportunity to add some of the useful lock information that we're parsing but not capturing today: https://github.com/anchore/syft/blob/main/syft/pkg/cataloger/javascript/parse_package_lock.go#L28-L29

Reply:
Can you elaborate on the license gathering? v1 lockfiles only use name and version, not path, so I'm a little confused as to what could've broken here. Also a good opportunity to write a new test

Reply:
Yeah, I should have noted this earlier -- there's a secondary lookup for licenses from `node_modules`: https://github.com/anchore/syft/blob/main/syft/pkg/cataloger/javascript/cataloger.go#L38 this has the assumption that the package name is found like `node_modules/<package-name>`, but in the case of a nested module like `node_modules/<different-package>/node_modules/<package-name>` we should retain that information so the license lookup is accurate

Reply:
Ah okay, I think I follow now. Would the new metadata type be correct, or is the Locations data better? I'm a little confused on what Locations are